### PR TITLE
Aws iam enhancement

### PR DESCRIPTION
**Pull request by YuvHayun**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [CIAC-6720](https://jira-hq.paloaltonetworks.local/browse/CIAC-6720)

## Description
AWS IAM enhancement

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

### CODE REVIEW

File: Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.py [hunk 1]
```diff
@@ -679,9 +679,14 @@ def get_role(args, aws_client):  # pragma: no cover
     })
 
     raw = json.loads(json.dumps(response['Role'], cls=DatetimeEncoder))
-    ec = {'AWS.IAM.Roles': raw}
+    raw["Tags"] = raw.get("Tags", [])
     human_readable = tableToMarkdown('AWS IAM Roles', data)
-    return_outputs(human_readable, ec)
+    return CommandResults(
+        outputs=createContext(raw),
+        outputs_prefix='AWS.IAM.Roles',
```
Comment (JudahSchwartz) [inline Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.py:686]: is this by design? the dot between aws.iam

**Review by JudahSchwartz — COMMENTED** [review]
looks great! lets tag docreview and schedule a demo

File: Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.yml [hunk 2]
```diff
@@ -529,7 +529,7 @@ script:
       description: The name of the IAM user that the new key will belong to.
       isArray: false
       name: userName
-      required: true
+      required: false
```
[the remark this answers is not in the document]
Comment (YuvHayun) [reply Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.yml:532]: Part of the design, in the last page.

File: Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.py (hunk 1 above)
Comment (YuvHayun) [reply Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.py:686]: Yes, this is the same output prefix for all commands in this integration.

File: Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.yml [hunk 3]
```diff
@@ -2304,7 +2310,171 @@ script:
...
+    name: aws-iam-tag-role
+  - arguments:
+    - default: false
+      description: The name of the IAM user to which you want to add tags.
+      isArray: false
+      name: userName
+      required: true
+      secret: false
+    - default: false
+      description: comma-separated list of Key:Value tags objects.
```
Comment (ShirleyDenkberg) [inline Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.yml:2407]:
```suggestion
      description: A comma-separated list of Key:Value tag objects.
```

File: Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.yml [hunk 4]
```diff
@@ -2304,7 +2310,171 @@ script:
...
+    name: aws-iam-tag-user
+  - arguments:
+    - default: false
+      description: The name of the IAM role to which you want to untag.
+      isArray: false
+      name: userName
+      required: true
+      secret: false
+    - default: false
+      description: comma-separated list of tags keys.
```
Comment (ShirleyDenkberg) [inline Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.yml:2424]:
```suggestion
      description: A comma-separated list of tag keys.
```

File: Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.yml [hunk 5]
```diff
@@ -2304,7 +2310,171 @@ script:
...
+    name: aws-iam-untag-user
+  - arguments:
+    - default: false
+      description: The name of the IAM role to which you want to untag.
+      isArray: false
+      name: roleName
+      required: true
+      secret: false
+    - default: false
+      description: comma-separated list of tags keys.
```
Comment (ShirleyDenkberg) [inline Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.yml:2441]:
```suggestion
      description: A comma-separated list of tag keys.
```

File: Packs/AWS-IAM/Integrations/AWS-IAM/README.md [hunk 6]
```diff
@@ -1690,3 +1692,217 @@ Lists all managed policies that are attached to the specified IAM user.
...
+#### Base Command
+
+`aws-iam-untag-user`
+
+#### Input
+
+| **Argument Name** | **Description** | **Required** |
+| --- | --- | --- |
+| userName | The name of the IAM role to which you want to untag. | Required | 
+| tagKeys | comma-separated list of tags keys. | Required |
```
Comment (ShirleyDenkberg) [inline Packs/AWS-IAM/Integrations/AWS-IAM/README.md:1837]:
```suggestion
| tagKeys | A comma-separated list of tag keys. | Required | 
```

File: Packs/AWS-IAM/Integrations/AWS-IAM/README.md [hunk 7]
```diff
@@ -1690,3 +1692,217 @@ Lists all managed policies that are attached to the specified IAM user.
...
+#### Base Command
+
+`aws-iam-untag-role`
+
+#### Input
+
+| **Argument Name** | **Description** | **Required** |
+| --- | --- | --- |
+| roleName | The name of the IAM role to which you want to untag. | Required | 
+| tagKeys | comma-separated list of tags keys. | Required |
```
Comment (ShirleyDenkberg) [inline Packs/AWS-IAM/Integrations/AWS-IAM/README.md:1864]:
```suggestion
| tagKeys | A comma-separated list of tag keys. | Required | 
```

File: Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.yml [hunk 8]
```diff
@@ -2304,7 +2310,171 @@ script:
...
+    name: aws-iam-untag-role
+  - arguments:
+    - default: false
+      description: The identifier of an access key.
+      isArray: false
+      name: accessKeyId
+      required: true
+      secret: false
+    deprecated: false
+    description: Retrieves information about when the specified access key was last used. The information includes the date and time of last use, along with the AWS service and Region that were specified in the last request made with that key.
```
Comment (ShirleyDenkberg) [inline Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.yml:2458]:
```suggestion
    description: Retrieves information about when the specified access key was last used. The information includes the date and time of last use, along with the AWS service and region that were specified in the last request made with that key.
```

File: Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.yml [hunk 9]
```diff
@@ -2304,7 +2310,171 @@ script:
...
+    name: aws-iam-get-access-key-last-used
+    outputs:
+    - contextPath: AWS.IAM.AccessKey.ID
+      description: The access key ID.
+      type: string
+    - contextPath: AWS.IAM.AccessKey.UserName
+      description: The username owning the given access key.
+      type: string
+    - contextPath: AWS.IAM.AccessKey.LastUsedServiceName
+      description: The name of the service last used the given access key.
```
Comment (ShirleyDenkberg) [inline Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.yml:2469]:
```suggestion
      description: The name of the service that last used the given access key.
```

File: Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.yml [hunk 10]
```diff
@@ -2304,7 +2310,171 @@ script:
...
+      type: string
+    - contextPath: AWS.IAM.AccessKey.LastUsedRegion
+      description: The name of the region where the given access key was last used in.
+      type: string
+    - contextPath: AWS.IAM.AccessKey.LastUsedDate
+      description: The date when the given access key was last used at.
```
Comment (ShirleyDenkberg) [inline Packs/AWS-IAM/Integrations/AWS-IAM/AWS-IAM.yml:2475]:
```suggestion
      description: The date when the given access key was last used.
```

File: Packs/AWS-IAM/Integrations/AWS-IAM/README.md [hunk 11]
```diff
@@ -1690,3 +1692,217 @@ Lists all managed policies that are attached to the specified IAM user.
...
+#### Base Command
+
+`aws-iam-tag-user`
+
+#### Input
+
+| **Argument Name** | **Description** | **Required** |
+| --- | --- | --- |
+| userName | The name of the IAM user to which you want to add tags. | Required | 
+| tags | comma-separated list of Key:Value tags objects. | Required |
```
Comment (ShirleyDenkberg) [inline Packs/AWS-IAM/Integrations/AWS-IAM/README.md:1810]:
```suggestion
| tags | A comma-separated list of Key:Value tag objects. | Required | 
```

File: Packs/AWS-IAM/Integrations/AWS-IAM/README.md [hunk 12]
```diff
@@ -1690,3 +1692,217 @@ Lists all managed policies that are attached to the specified IAM user.
...
+#### Base Command
+
+`aws-iam-tag-role`
+
+#### Input
+
+| **Argument Name** | **Description** | **Required** |
+| --- | --- | --- |
+| roleName | The name of the IAM role to which you want to add tags. | Required | 
+| tags | comma-separated list of Key:Value tags objects. | Required |
```
Comment (ShirleyDenkberg) [inline Packs/AWS-IAM/Integrations/AWS-IAM/README.md:1783]:
```suggestion
| tags | A comma-separated list of Key:Value tag objects. | Required | 
```

File: Packs/AWS-IAM/Integrations/AWS-IAM/README.md [hunk 13]
```diff
@@ -1690,3 +1692,217 @@ Lists all managed policies that are attached to the specified IAM user.
...
+
+### Untagged the following tags from role {role_name}
+|Removed keys|
+|---|
+| Key1 |
+
+### aws-iam-get-access-key-last-used
+
+***
+Retrieves information about when the specified access key was last used. The information includes the date and time of last use, along with the AWS service and Region that were specified in the last request made with that key.
```
Comment (ShirleyDenkberg) [inline Packs/AWS-IAM/Integrations/AWS-IAM/README.md:1880]:
```suggestion
Retrieves information about when the specified access key was last used. The information includes the date and time of last use, along with the AWS service and region that were specified in the last request made with that key.
```

File: Packs/AWS-IAM/Integrations/AWS-IAM/README.md [hunk 14]
```diff
@@ -1690,3 +1692,217 @@ Lists all managed policies that are attached to the specified IAM user.
...
+| --- | --- | --- |
+| accessKeyId | The identifier of an access key. | Required | 
+
+#### Context Output
+
+| **Path** | **Type** | **Description** |
+| --- | --- | --- |
+| AWS.IAM.AccessKey.ID | string | The access key ID. | 
+| AWS.IAM.AccessKey.UserName | string | The username owning the given access key. | 
+| AWS.IAM.AccessKey.LastUsedServiceName | string | The name of the service last used the given access key. |
```
Comment (ShirleyDenkberg) [inline Packs/AWS-IAM/Integrations/AWS-IAM/README.md:1898]:
```suggestion
| AWS.IAM.AccessKey.LastUsedServiceName | string | The name of the service that last used the given access key. | 
```

File: Packs/AWS-IAM/Integrations/AWS-IAM/README.md [hunk 15]
```diff
@@ -1690,3 +1692,217 @@ Lists all managed policies that are attached to the specified IAM user.
...
+
+#### Context Output
+
+| **Path** | **Type** | **Description** |
+| --- | --- | --- |
+| AWS.IAM.AccessKey.ID | string | The access key ID. | 
+| AWS.IAM.AccessKey.UserName | string | The username owning the given access key. | 
+| AWS.IAM.AccessKey.LastUsedServiceName | string | The name of the service last used the given access key. | 
+| AWS.IAM.AccessKey.LastUsedRegion | string | The name of the region where the given access key was last used in. | 
+| AWS.IAM.AccessKey.LastUsedDate | string | The date when the given access key was last used at. |
```
Comment (ShirleyDenkberg) [inline Packs/AWS-IAM/Integrations/AWS-IAM/README.md:1900]:
```suggestion
| AWS.IAM.AccessKey.LastUsedDate | string | The date when the given access key was last used. | 
```

File: Packs/AWS-IAM/ReleaseNotes/1_1_32.md [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+  - ***aws-iam-put-role-policy***
+  - ***aws-iam-put-user-policy***
+  - ***aws-iam-put-group-policy***
+  - ***aws-iam-tag-role***
+  - ***aws-iam-tag-user***
+  - ***aws-iam-untag-user***
+  - ***aws-iam-untag-role***
+  - ***aws-iam-get-access-key-last-used***
+  For more information, refer to the integration readme.
+- Note: the **username** argument for ***aws-iam-update-access-key***, ***aws-iam-delete-access-key***, ***aws-iam-create-access-key*** commands is not mandatory anymore.
```
Comment (ShirleyDenkberg) [inline Packs/AWS-IAM/ReleaseNotes/1_1_32.md:13]:
```suggestion
- Note: the *username* argument for the following commands is not mandatory anymore:L
    - ***aws-iam-update-access-key***
    - ***aws-iam-delete-access-key***
    - ***aws-iam-create-access-key*** 
```

**Comment by ShirleyDenkberg** [comment]
@JudahSchwartz Doc review completed.

**Review by JudahSchwartz — APPROVED** [review]
Awesomesauce!

**Comment by YuvHayun** [comment]
@dantavori need force merge on this PR, missing 0.02% on coverage (added tests to all the commands I added).

**Pull request merged**